server/utils/stigLibrary.ts

```diff
@@ -0,0 +1,291 @@
+import * as fs from "fs";
+import * as path from "path";
+import AdmZip from "adm-zip";
+import { DateTime } from "luxon";
+import { Op, UniqueConstraintError } from "sequelize";
+import {
+  Assessment,
+  AssessmentItem,
+  StigData,
+  StigLibrary,
+  System,
+  Stig,
+  StigAlias,
+} from "../../db/models";
+import { hashFile } from "./hash";
+import { createBlankAssessment } from "./assessments";
+
+export type ProcessLibraryResults = {
+  stigProcessed: number;
+  newStigCount: number;
+  updatedStigCount: number;
+  unchangedStigCount: number;
+  xmlsExtracted: number;
+};
+
+export const migrateBoundary = async (
+  boundaryId: number,
+  newStigLibraryId: number,
+): Promise<{ results: string }> => {
+  try {
+    const boundarySystems = await System.findAll({ where: { BoundaryId: boundaryId } });
+    for (const system of boundarySystems) {
+      const currentAssessments = await Assessment.findAll({
+        where: {
+          SystemId: system.id,
+          succeededByAssessmentId: { [Op.is]: null },
+        },
+      });
+      for (const assessment of currentAssessments) {
+        const oldStig = await Stig.findOne({ where: { id: assessment.StigId } });
+        const newStig = await Stig.findOne({
+          where: { stigid: oldStig.stigid },
+          include: [
+            {
+              model: StigLibrary,
+              where: { id: newStigLibraryId },
+              required: true,
+            },
+          ],
+        });
+
+        await system.removeStig(oldStig);
+        await system.addStig(newStig);
+
+        const newAssessment = await createBlankAssessment(assessment.SystemId, newStig.id);
+
+        assessment.succeededByAssessmentId = newAssessment.id;
+        await assessment.save();
+
+        const oldChecks = await AssessmentItem.findAll({
+          where: { AssessmentId: assessment.id },
+          include: [{ model: StigData, attributes: ["vuln_num", "rule_id"] }],
+        });
+        const newChecks = await AssessmentItem.findAll({
+          where: { AssessmentId: newAssessment.id },
+          include: [{ model: StigData, attributes: ["vuln_num", "rule_id"] }],
+        });
+
+        for (const check of newChecks) {
+          const matchingCheck = findMatch(check, oldChecks);
```

Comment:
maybe explore doing preprocessing here instead of doing a find in each iteration here.  it'd be something that'd return the full union set as well as the set that remains unique to each assessment.  you can iterate over all of that in one go then.